src/goto-instrument/value_set_fi_fp_removal.cpp

```diff
@@ -0,0 +1,249 @@
+/*******************************************************************\
+
+Module: value_set_fi_Fp_removal
+
+Author: Daniel Kroening, kroening@kroening.com
+
+\*******************************************************************/
+
+#include "value_set_fi_fp_removal.h"
+
+#include <goto-programs/goto_model.h>
+#include <goto-programs/remove_function_pointers.h>
+
+#include <pointer-analysis/value_set_analysis_fi.h>
+
+#include <util/base_type.h>
+#include <util/c_types.h>
+#include <util/expanding_vector.h>
+#include <util/fresh_symbol.h>
+#include <util/message.h>
+#include <util/namespace.h>
+#include <util/std_code.h>
+#include <util/union_find.h>
+
+#ifdef USE_STD_STRING
+#  include <util/dstring.h>
+#endif
+
+
+void fix_argument_types(
+  code_function_callt &function_call,
+  const namespacet &ns)
+{
+  const code_typet &code_type =
+    to_code_type(ns.follow(function_call.function().type()));
+
+  const code_typet::parameterst &function_parameters = code_type.parameters();
+
+  code_function_callt::argumentst &call_arguments = function_call.arguments();
+
+  for(std::size_t i = 0; i < function_parameters.size(); i++)
+  {
+    // casting pointers might result in more arguments than function parameters
+    if(i < call_arguments.size())
+    {
+      if(call_arguments[i].type() != function_parameters[i].type())
```

Comment:
Can remove this if statement now (conditional_cast does this)

Reply:
To be clear, I mean that `conditional_cast(a, b)` expands to `return a.type() != b ? typecast(a, b) : a`, so there's no need to surround it in an additional `if(a.type() != b)`